### ISSUE TITLE
Bad quantized model compilation time for RISC-V target

### ISSUE THREAD

**Issue by hanhanW** [google/iree#8431]
This is the bug mainly for tracking compilation time regression on RISC-V targets.

We used to compile the (quantized) person_detection.tflite model at 3 seconds. After landing https://github.com/google/iree/pull/8409, the quantized_matmul ops are lowered to matmul ops which kicks in matmul vectorization. It takes 13 min to compile the model.

With the fixes in https://github.com/google/iree/issues/8210, we're able to compile the model in 150 seconds (which is still bad).   After profiling the pass-timing, I found that most of the time spent in HAL-related passes.

```
===-------------------------------------------------------------------------===
                         ... Execution time report ...
===-------------------------------------------------------------------------===
  Total Execution Time: 140.5211 seconds

  ----User Time----  ----Wall Time----  ----Name----
  297.5860 (196.1%)  276.3979 (196.7%)  'hal.executable' Pipeline
  151.7494 (100.0%)  140.5211 (100.0%)  root
  149.1610 ( 98.3%)  138.2025 ( 98.3%)  Pipeline Collection : ['hal.executable']
  137.6607 ( 90.7%)  137.6607 ( 98.0%)  mlir::iree_compiler::IREE::HAL::SerializeExecutablesPass
  137.6586 ( 90.7%)  137.6586 ( 98.0%)  mlir::iree_compiler::IREE::HAL::SerializeTargetExecutablesPass
   25.8591 ( 17.0%)   25.8591 ( 18.4%)  'hal.executable.variant' Pipeline
   11.0800 (  7.3%)   11.0800 (  7.9%)  Pipeline Collection : ['hal.executable.variant']
   10.9722 (  7.2%)   10.9722 (  7.8%)  mlir::iree_compiler::IREE::HAL::TranslateTargetExecutableVariantsPass
    8.1665 (  5.4%)    8.1665 (  5.8%)  Pipeline Collection : ['builtin.module']
    6.4248 (  4.2%)    6.4248 (  4.6%)  LLVMCPULowerExecutableTarget
    5.1236 (  3.4%)    5.1236 (  3.6%)  'builtin.module' Pipeline
    3.3519 (  2.2%)    3.3519 (  2.4%)  Pipeline Collection : ['builtin.func']
    1.3615 (  0.9%)    1.3615 (  1.0%)  'builtin.func' Pipeline
    1.2922 (  0.9%)    1.2502 (  0.9%)  Canonicalizer
    1.2305 (  0.8%)    1.2305 (  0.9%)  ConvertToLLVM
    0.7574 (  0.5%)    0.7574 (  0.5%)  mlir::iree_compiler::IREE::Util::(anonymous namespace)::FoldGlobalsPass
   11.1462 (  7.3%)    0.5405 (  0.4%)  mlir::iree_compiler::IREE::HAL::TranslateExecutablesPass
    0.3919 (  0.3%)    0.3919 (  0.3%)  mlir::iree_compiler::IREE::Util::ApplyPatternsPass
    0.3122 (  0.2%)    0.2910 (  0.2%)  CSE
```

We don't invest much effort for quantized vectorization models, so I'm not surprised seeing such issue. A temp solution could be falling it back to non-vectorization pipeline if the target is RISC-V.

To repro:

1. Download tflite file: https://github.com/google/iree/blob/main/integrations/tensorflow/test/python/iree_tfl_tests/person_detect_test.py#L14
2. import it into MLIR with `iree-import-tosa`
3. Run the translation targeting on RISC-V: `iree-translate -iree-mlir-to-vm-bytecode-module --iree-hal-target-backends=dylib-llvm-aot --iree-llvm-link-embedded=tru
e -iree-input-type=tosa ../tosa.mlir -o /tmp/a.vmfb -iree-llvm-target-triple=riscv64 -iree-llvm-target-cpu=generic-rv64 -iree-llvm-target-abi=lp64d -
iree-llvm-target-cpu-features=+m,+a,+f,+d,+c`

**Comment by hcindyl** [comment]
We need to enable vectorization option for the RISCV target for other design explorations, so please don't apply fallback for this target. 150s compilation is not ideal, but probably is something we can tolerate in order to explore more into the workload/artifact.

It will be nice though, to have a specific flag to turn on/off vectorization so we can flip the flag and speed up the build time for some workload.

**Comment by hcindyl** [comment]
Could you please provide the CLI or the flags to produce the execution time report? I suspect other RISC-V configs on the same model would be even worse.

RV32 RVV static library:
```
iree-translate -iree-mlir-to-vm-bytecode-module --iree-hal-target-backends=dylib-llvm-aot \
--iree-llvm-target-triple=riscv32-pc-linux-elf -iree-llvm-target-cpu=generic-rv32 -iree-llvm-target-abi=ilp32\
--iree-llvm-target-cpu-features=+m,+a,+f,+zvl512b,zve32x \
--riscv-v-vector-bits-min=512 --riscv-v-fixed-length-vector-lmul-max=8 \
--iree-llvm-link-embedded=false --iree-llvm-link-static --iree-llvm-system-linker-path=lld  \
--iree-llvm-static-library-output-path=/tmp/a.o
--iree-input-type=tosa ../tosa.mlir -o /tmp/a.vmfb 
```

I test locally without the fix and the compilation hangs for >110 minutes and counting. RV64 RVV embedded library takes 15+ minutes to run per CI log.

**Comment by hanhanW** [comment]
You can use `-pass-timing` flag to get the execution time report.

Or use Tracy and disable mlir threading to profile the compilation time.

**Comment by hanhanW** [comment]
I'm not pretty sure how we select tiling sizes in this case. The matmul stuff goes with TileFuseAndVectorize may generates bunch of vectors if the tiling sizes is not set properly. (Because there are vector unrolling stuff kicked in).

**Comment by hcindyl** [comment]
Hi @hanhanW , is the fix landed in IREE? I am happy to provide more information for the RISC-V targets but I need to know where to cherry-pick/apply the change.

**Comment by hanhanW** [comment]
The fix is still in-reivew, I'm addressing the comments, and I will let you know once I land it.

https://reviews.llvm.org/D120760

**Comment by hanhanW** [comment]
The fix was landed. We'll need integrations to kick them in IREE. I'm the integrator for this week and I'll probably start one tomorrow. https://github.com/llvm/llvm-project/commit/1538bd518cd236f4321695e9c5f0dd24601db366

**Comment by hcindyl** [comment]
Awesome! Looking forward to it.

**Comment by hanhanW** [comment]
@hcindyl The patch is landed in IREE within https://github.com/google/iree/commit/6c182f591dd666222589115a5edf0aa5e556ed2b

**Comment by benvanik** [comment]
Thanks hanhan!

**Comment by hcindyl** [comment]
Thanks. The presubmit log shows a better compilation time for RV64 embedded targets (5min -> 3min) https://source.cloud.google.com/results/invocations/62d9daa5-78a2-4023-be66-a9338bfcac6d/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Friscv64%2Fpresubmit/log

We will test the rv32 static library targets and report back here.

**Comment by hcindyl** [comment]
> You can use `-pass-timing` flag to get the execution time report.
> 
> Or use Tracy and disable mlir threading to profile the compilation time.

I can't find the `--pass-timing` flag in `iree-translate --helpall` What's the proper flag to generate the execution report, `--mlir-timing`?

**Comment by lundong** [comment]
@hanhanW 

The new fix does not seem to improve for RV32. The compilation time is still extremely long.

I tested RV32 (both static library and embedded dylib; results are similar) for the person_detection model on my machine. For RV32 with RVV, the compilation time is > 20 min. For RV32 without RVV, the compilation time is > 5 min.

Caveat: the above number is machine-dependent and have run-by-run variations.

**Comment by mariecwhite** [comment]
I'm also running into long compilation times for quantized models for x86 and ARM. Some example MLIRs [attached here](https://github.com/google/iree/files/8229497/mobilebert_deeplab.zip) for quantized MobileBert and Deeplab. My build kite runs don't complete, even after [11 hours](https://buildkite.com/iree/iree-benchmark/builds/2224).

**Comment by lundong** [comment]
@hanhanW 
Flags I used:
 -iree-mlir-to-vm-bytecode-module
 -iree-hal-target-backends=dylib-llvm-aot
 -iree-llvm-target-triple=riscv32-pc-linux-elf
 -iree-llvm-target-cpu=generic-rv32
 -iree-llvm-target-cpu-features=+m,+f (=+m,+f,+zvl512b,+zve32x for RVV ON)
 -iree-llvm-target-abi=ilp32
 -iree-llvm-link-embedded=true
 -iree-llvm-debug-symbols=false
 -iree-input-type=tosa
 -riscv-v-vector-bits-min=512
 -riscv-v-fixed-length-vector-lmul-max=8
 -mlir-timing

**Comment by hanhanW** [comment]
I dig into details and make it much better. It takes 45 secs when targeting rv32 w/o RVV. And it takes 2.5 mins when targeting rv32 w/ RVV. I'll send out the PR soon.

This is because we have bad configurations for RISC-V targets. They are all using ARM CPU configurations (which is mainly targeting Pixel 4).

Re Maries' issue, I think they are different issues. I'll file another issue for tracking that.

**Comment by lundong** [comment]
@hanhanW Thanks. I will give a try and let you know how it goes.

Also, you mentioned that the above RV32 configurations were bad. Can you please suggest a set of improved configurations?

**Comment by hanhanW** [comment]
The configurations I mentioned are more about choosing the codegen pipeline and tiling sizes. IREE determines the codegen pipeline based on anchor ops and target platform, and select tiling sizes heuristically with native vector sizes. We have fine configurations now. They are not best tuned yet, but they are reasonable configurations to me.

**Comment by lundong** [comment]
@hanhanW With your latest fix (in candidate-20220316.78), the compilation time for rv32 targets is indeed improved greatly and may be okay now. Thanks for your work!

However, the main issue now is that the compiled code (.o) size is significantly larger. Just give an example, for the person_detection model, the .o file size is 215KB in candidate-20220225.59, but it is 3.2MB in candidate-20220316.78. I have included those two files for your reference. Can you please look into it?
[person_detection.tar.gz](https://github.com/google/iree/files/8274861/person_detection.tar.gz)

**Comment by hanhanW** [comment]
I'm not surprised that the binary size increases, but it's significantly larger. We'll have to look into it. However, I think it's not the top priority on my plate. I think we should sort out what we wanna achieve, set up better configurations for performance, then we can optimize the binary sizes.

This issue is mostly for tracking compilation time issue for RISC-V. Moving further discussion to https://github.com/google/iree/issues/8568
